### PR TITLE
use material property for barrier height in phase field classes,

### LINKED ISSUE #16778: Phase-field barrier height could use material property
## Reason
For ease-of-use and consistency with other phase-field parameters (mobility, gradient coefficients), the barrier height could be stored in a material property and passed to objects that need it rather than setting the value manually. This is already done in `PolynomialFreeEnergy`.

## Design
Store the barrier height as a Material and have other objects read it.

## Impact
More flexibility. Existing method of setting the value with a floating point value would still work.

### PR DESCRIPTION
**Pull request by jessecarterMOOSE**
Went through and found several places where the phase-field barrier height value was read in using a Real parameter and changed it to a Real material property. Tests do not require a change since `w = 2` still works.

closes #16778

### CODE REVIEW

**Comment by jessecarterMOOSE** [comment]
@dschwen here is this. I'm not quite sure what to do about [`CrossTermBarrierFunctionMaterial`](https://github.com/idaholab/moose/blob/next/modules/phase_field/src/materials/CrossTermBarrierFunctionMaterial.C).

**Comment by dschwen** [comment]
> @dschwen here is this. I'm not quite sure what to do about [`CrossTermBarrierFunctionMaterial`](https://github.com/idaholab/moose/blob/next/modules/phase_field/src/materials/CrossTermBarrierFunctionMaterial.C).

Nothing for now. That's ok.

**Review by dschwen — CHANGES_REQUESTED** [review]
This needs a test.

**Comment by jessecarterMOOSE** [comment]
Added a new test using a kks example though i probably could have gotten away with adding the functionality to an existing test. 

What is with the failure?

File: modules/phase_field/test/tests/KKS_system/kks_barrier_height_material.i [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    type = BarrierFunctionMaterial
+    g_order = SIMPLE
+    eta = eta
+  [../]
+
+  # constant properties
+  [./constants]
+    type = GenericConstantMaterial
+    prop_names  = 'M   L   kappa barrier_height'
+    prop_values = '0.7 0.7 0.4   0.4'
```
Comment (dschwen) [inline modules/phase_field/test/tests/KKS_system/kks_barrier_height_material.i:128]: This is not a good test. We need to test the _new_ aspect of this capability, which is a spatially varying barrier height.

Comment (jessecarterMOOSE) [reply modules/phase_field/test/tests/KKS_system/kks_barrier_height_material.i:128]: Makes sense, though I don't have a use case for that yet. The test would just be some arbitrary variation.

Comment (jessecarterMOOSE) [reply modules/phase_field/test/tests/KKS_system/kks_barrier_height_material.i]: OK, added a test with a barrier height that depends on concentration.

**Comment by jessecarterMOOSE** [comment]
@dschwen this should be ready for another review

**Comment by jessecarterMOOSE** [comment]
@dschwen any further thoughts on this?

**Comment by dschwen** [comment]
@jessecarterMOOSE I think this is not a variationally correct formulation! In `DerivativeTwoPhaseMaterial::computeDF` you are ignoring potential derivatives of the barrier height function with respect to other nonlinear variables in your problem. You get a correct Jacobian, but your residual is already incorrect (first derivative of the free energy density goes into that).

**Comment by jessecarterMOOSE** [comment]
@dschwen Is that because previously the value was considered a constant and now it can vary spatially? Do DerivativeParsedMaterials not handle that?

I really only made this change for flexibility and still intended on using a constant barrier height.

**Comment by dschwen** [comment]
I'd suggest just using input file variables, e.g.

W = ${barrier_height}

**Comment by dschwen** [comment]
Derivative parsed materials can supply the necessary derivatives but you have to actually use them!

**Comment by dschwen** [comment]
I'm closing this as this would introduce a potentially inconsistent formulation. I suggest using input file brace expansion variables.

**Pull request closed without merging**